# Analyze Product Sentiment

In [1]:
import turicreate

# Read product review data

In [3]:
products = turicreate.SFrame('amazon_baby.sframe')
# products = turicreate.SFrame('~/data/amazon_baby.sframe')

# Explore data

In [4]:
products

Columns:
	name	str
	review	str
	rating	float

Rows: 183531

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |  3.0   |
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
+-------------------------------+-------------------------------+--------+
[183531 rows x 3 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [6]:
products.groupby('name',operations={'count':turicreate.aggregate.COUNT()}).sort('count',ascending=False)

Columns:
	name	str
	count	int

Rows: 32419

Data:
+--------------------------------+-------+
|              name              | count |
+--------------------------------+-------+
| Vulli Sophie the Giraffe T...  |  785  |
| Simple Wishes Hands-Free B...  |  562  |
| Infant Optics DXR-5 2.4 GH...  |  561  |
| Baby Einstein Take Along Tunes |  547  |
| Cloud b Twilight Constella...  |  520  |
| Fisher-Price Booster Seat,...  |  489  |
| Fisher-Price Rainforest Ju...  |  450  |
| Graco Nautilus 3-in-1 Car ...  |  419  |
| Leachco Snoogle Total Body...  |  388  |
| Regalo Easy Step Walk Thru...  |  374  |
+--------------------------------+-------+
[32419 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

# Examine the reivews for the most-reviewed product

In [7]:
giraffe_reviews = products[products['name']=='Vulli Sophie the Giraffe Teether']

In [8]:
giraffe_reviews

Columns:
	name	str
	review	str
	rating	float

Rows: Unknown

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
| Vulli Sophie the Giraffe T... | He likes chewing on all th... |  5.0   |
| Vulli Sophie the Giraffe T... | My son loves this toy and ... |  5.0   |
| Vulli Sophie the Giraffe T... | There really should be a l... |  1.0   |
| Vulli Sophie the Giraffe T... | All the moms in my moms' g... |  5.0   |
| Vulli Sophie the Giraffe T... | I was a little skeptical o... |  5.0   |
| Vulli Sophie the Giraffe T... | I have been reading about ... |  5.0   |
| Vulli Sophie the Giraffe T... | My neice loves her sophie ... |  5.0   |
| Vulli Sophie the Giraffe T... | What a friendly face!  And... |  5.0   |
| Vulli Sophie the Giraffe T... | We got this just for my so... |  5.0   |
| Vulli Sophie the Giraffe T... | My baby seems to like this... |  3.0   |
+-------------------------------+-------------------------------+--------+
[? rows x 3 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

In [9]:
len(giraffe_reviews)

785

In [10]:
giraffe_reviews['rating'].show()

Materializing SArray

# Building a sentiment classifier

## Build word count vectors

In [11]:
products['word_count'] = turicreate.text_analytics.count_words(products['review'])

In [12]:
products

Columns:
	name	str
	review	str
	rating	float
	word_count	dict

Rows: 183531

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |  3.0   |
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+
|           word_count          |
+-------------------------------+
| {'handles': 1.0, 'strippin... |
| {'recommend': 1.0, 'highly... |
| {'quilt': 1.0, 'of': 1.0, ... |
| {'tool': 1.0, 'clever': 1.... |
| {'rock': 1.0, 'many': 1.0,... |
| {'thumb': 1.0, 'or': 1.0, ... |
| {'for': 1.0, 'barnes': 1.0... |
| {'right': 1.0, 'because': ... |
| {'like': 1.0, 'and': 1.0, ... |
| {'in': 1.0, 'pages': 1.0, ... |
+-------------------------------+
[183531 rows x 4 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [31]:
products['name'].show()

Materializing SArray

# Define what is positive and negative sentiment

In [13]:
products['rating'].show()

Materializing SArray

In [14]:
#ignore all 3*  reviews
products = products[products['rating']!= 3]

In [15]:
#positive sentiment = 4-star or 5-star reviews
products['sentiment'] = products['rating'] >= 4

In [16]:
products

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int

Rows: 166752

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
| Baby Tracker&reg; - Daily ... | I love this journal and ou... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+
|           word_count          | sentiment |
+-------------------------------+-----------+
| {'recommend': 1.0, 'highly... |     1     |
| {'quilt': 1.0, 'of': 1.0, ... |     1     |
| {'tool': 1.0, 'clever': 1.... |     1     |
| {'rock': 1.0, 'many': 1.0,... |     1     |
| {'thumb': 1.0, 'or': 1.0, ... |     1     |
| {'for': 1.0, 'barnes': 1.0... |     1     |
| {'right': 1.0, 'because': ... |     1     |
| {'like': 1.0, 'and': 1.0, ... |     1     |
| {'in': 1.0, 'pages': 1.0, ... |     1     |
| {'tracker': 1.0, 'now': 1.... |     1     |
+-------------------------------+-----------+
[166752 rows x 5 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [17]:
products['sentiment'].show()

Materializing SArray

# Train our sentiment classifier

In [18]:
train_data,test_data = products.random_split(.8,seed=0)

In [19]:
sentiment_model = turicreate.logistic_classifier.create(train_data,target='sentiment', features=['word_count'], validation_set=test_data)

Logistic regression:

--------------------------------------------------------

Number of examples          : 133448

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 57356

Number of coefficients      : 57357

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 4        | 0.250000  | 1.757196     | 0.841421          | 0.840019            |

| 1         | 9        | 3.250000  | 2.558561     | 0.931359          | 0.911362            |

| 2         | 11       | 2.669978  | 3.040334     | 0.938650          | 0.916466            |

| 3         | 12       | 2.669978  | 3.350289     | 0.927417          | 0.901814            |

| 4         | 14       | 1.326072  | 3.718972     | 0.945200          | 0.918088            |

| 9         | 20       | 1.326072  | 5.494454     | 0.977654          | 0.917698            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

In [33]:
results = sentiment_model.evaluate(products)
results

{'accuracy': 0.9656795720591057,
 'auc': 0.983638617259335,
 'confusion_matrix': Columns:
 	target_label	int
 	predicted_label	int
 	count	int
 
 Rows: 4
 
 Data:
 +--------------+-----------------+--------+
 | target_label | predicted_label | count  |
 +--------------+-----------------+--------+
 |      1       |        0        |  2569  |
 |      0       |        1        |  3154  |
 |      0       |        0        | 23339  |
 |      1       |        1        | 137690 |
 +--------------+-----------------+--------+
 [4 rows x 3 columns],
 'f1_score': 0.9796409145402218,
 'log_loss': 0.11725006745036073,
 'precision': 0.9776064298088666,
 'recall': 0.9816838848130958,
 'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 100001
 
 Data:
 +-----------+--------------------+--------------------+--------+-------+
 | threshold |        fpr         |        tpr         |   p    |   n   |
 +-----------+--------------------+--------------------+--------+---

# Apply the sentiment classifier to better understand the Giraffe reviews

In [20]:
products['predicted_sentiment'] = sentiment_model.predict(products, output_type = 'probability')

In [21]:
products

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	predicted_sentiment	float

Rows: 166752

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
| Baby Tracker&reg; - Daily ... | I love this journal and ou... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+---------------------+
|           word_count          | sentiment | predicted_sentiment |
+-------------------------------+-----------+---------------------+
| {'recommend': 1.0, 'highly... |     1     |  0.9997307390047108 |
| {'quilt': 1.0, 'of': 1.0, ... |     1     |  0.9985083368316828 |
| {'tool': 1.0, 'clever': 1.... |     1     |  0.9997489042499907 |
| {'rock': 1.0, 'many': 1.0,... |     1     |  0.9999916625399978 |
| {'thumb': 1.0, 'or': 1.0, ... |     1     |  0.9999999514462167 |
| {'for': 1.0, 'barnes': 1.0... |     1     |  0.9999146735569922 |
| {'right': 1.0, 'because': ... |     1     |  0.9999916615904655 |
| {'like': 1.0, 'and': 1.0, ... |     1     |  0.999993884359401  |
| {'in': 1.0, 'pages': 1.0, ... |     1     |  0.9961247617005238 |
| {'tracker': 1.0, 'now': 1.... |     1     |  0.9999999920460633 |
+-------------------------------+-----------+---------------------+
[166752 rows x 6 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [22]:
giraffe_reviews = products[products['name']== 'Vulli Sophie the Giraffe Teether']

In [23]:
giraffe_reviews

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	predicted_sentiment	float

Rows: Unknown

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
| Vulli Sophie the Giraffe T... | He likes chewing on all th... |  5.0   |
| Vulli Sophie the Giraffe T... | My son loves this toy and ... |  5.0   |
| Vulli Sophie the Giraffe T... | There really should be a l... |  1.0   |
| Vulli Sophie the Giraffe T... | All the moms in my moms' g... |  5.0   |
| Vulli Sophie the Giraffe T... | I was a little skeptical o... |  5.0   |
| Vulli Sophie the Giraffe T... | I have been reading about ... |  5.0   |
| Vulli Sophie the Giraffe T... | My neice loves her sophie ... |  5.0   |
| Vulli Sophie the Giraffe T... | What a friendly face!  And... |  5.0   |
| Vulli Sophie the Giraffe T... | We got this just for my so... |  5.0   |
| Vulli Sophie the Giraffe T... | This product is without a ... |  5.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+---------------------+
|           word_count          | sentiment | predicted_sentiment |
+-------------------------------+-----------+---------------------+
| {'purchase': 1.0, 'teethin... |     1     |  0.9993655365682544 |
| {'a': 1.0, 'is': 1.0, 'whe... |     1     |  0.9998633791689672 |
| {'made': 1.0, 'of': 1.0, '... |     0     |  0.2545268197491694 |
| {'another': 1.0, 'out': 1.... |     1     |  0.9165688083895887 |
| {'disappointed': 1.0, 'wil... |     1     |  0.6855768205778634 |
| {'late': 1.0, 'perfect': 1... |     1     |   0.99999994452112  |
| {'delight': 1.0, 'in': 1.0... |     1     |  0.997935118109276  |
| {'inside': 1.0, 'water': 1... |     1     |  0.9999745004834412 |
| {'its': 1.0, 'fine': 1.0, ... |     1     |  0.9460144428329694 |
| {'belongs': 1.0, 'dog': 1.... |     1     |  0.9999999721665869 |
+-------------------------------+-----------+---------------------+
[? rows x 6 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

# Sort the Giraffe reviews according to predicted sentiment

In [24]:
giraffe_reviews = giraffe_reviews.sort('predicted_sentiment', ascending=False)

In [25]:
giraffe_reviews

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	predicted_sentiment	float

Rows: 723

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
| Vulli Sophie the Giraffe T... | Sophie, oh Sophie, your ti... |  5.0   |
| Vulli Sophie the Giraffe T... | I'll be honest...I bought ... |  4.0   |
| Vulli Sophie the Giraffe T... | As a mother of 16month old... |  5.0   |
| Vulli Sophie the Giraffe T... | We got this little giraffe... |  5.0   |
| Vulli Sophie the Giraffe T... | As every mom knows, you al... |  5.0   |
| Vulli Sophie the Giraffe T... | My Mom-in-Law bought Sophi... |  5.0   |
| Vulli Sophie the Giraffe T... | My 4 month old son is teet... |  4.0   |
| Vulli Sophie the Giraffe T... | Let me just start off by a... |  5.0   |
| Vulli Sophie the Giraffe T... | I'm not sure why Sophie is... |  4.0   |
| Vulli Sophie the Giraffe T... | I admit, I didn't get Soph... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+---------------------+
|           word_count          | sentiment | predicted_sentiment |
+-------------------------------+-----------+---------------------+
| {'11': 1.0, 'prisrob': 1.0... |     1     |         1.0         |
| {'around': 1.0, 'explore':... |     1     |         1.0         |
| {'will': 1.0, '15months': ... |     1     |         1.0         |
| {'out': 1.0, 'would': 1.0,... |     1     |  0.9999999999998377 |
| {'whether': 1.0, 'neutral'... |     1     |  0.9999999999998284 |
| {'penny': 1.0, 'little': 1... |     1     |  0.9999999999997958 |
| {'worth': 1.0, 'works': 1.... |     1     |  0.9999999999994914 |
| {'question': 1.0, 'must': ... |     1     |  0.9999999999941254 |
| {'makers': 1.0, 'or': 1.0,... |     1     |  0.999999999987423  |
| {'dye': 1.0, 'of': 1.0, 'c... |     1     |  0.9999999999829475 |
+-------------------------------+-----------+---------------------+
[723 rows x 6 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [26]:
giraffe_reviews.tail()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	predicted_sentiment	float

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
| Vulli Sophie the Giraffe T... | My 7 1/2 month old grandba... |  1.0   |
| Vulli Sophie the Giraffe T... | I was so looking forward t... |  1.0   |
| Vulli Sophie the Giraffe T... | Received the product and s... |  1.0   |
| Vulli Sophie the Giraffe T... | I got one of these as a sh... |  1.0   |
| Vulli Sophie the Giraffe T... | When we first got it I tho... |  2.0   |
| Vulli Sophie the Giraffe T... | When I received this the p... |  1.0   |
| Vulli Sophie the Giraffe T... | "Sophie the Giraffe" has t... |  1.0   |
| Vulli Sophie the Giraffe T... | My son (now 2.5) LOVED his... |  1.0   |
| Vulli Sophie the Giraffe T... | I wanted to love this prod... |  1.0   |
| Vulli Sophie the Giraffe T... | This children's toy is nos... |  1.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+------------------------+
|           word_count          | sentiment |  predicted_sentiment   |
+-------------------------------+-----------+------------------------+
| {'sopie': 1.0, 'with': 1.0... |     0     | 1.911074019331925e-06  |
| {'amazon': 1.0, 'back': 1.... |     0     | 6.547801684255248e-07  |
| {'was': 1.0, 'as': 1.0, 's... |     0     | 3.864236396972233e-07  |
| {'typical': 1.0, 'paint': ... |     0     | 2.2378225152325034e-07 |
| {'mom': 1.0, 'teethers': 1... |     0     | 1.9353955279957842e-07 |
| {'risk': 1.0, 'but': 1.0, ... |     0     | 1.6900753979362622e-07 |
| {'ingest': 1.0, 'they': 1.... |     0     | 1.1748826795805357e-07 |
| {'exaggerated': 1.0, 'are'... |     0     | 5.364933426300608e-11  |
| {'risk': 1.0, 'ready': 1.0... |     0     | 3.1090352697495393e-13 |
| {'amazon': 1.0, 'from': 1.... |     0     | 4.119269638678007e-15  |
+-------------------------------+-----------+------------------------+
[10 rows x 6 columns]

## Show the most positive reviews

In [27]:
giraffe_reviews[0]['review']

"Sophie, oh Sophie, your time has come. My granddaughter, Violet is 5 months old and starting to teeth. What joy little Sophie brings to Violet. Sophie is made of a very pliable rubber that is sturdy but not tough. It is quite easy for Violet to twist Sophie into unheard of positions to get Sophie into her mouth. The little nose and hooves fit perfectly into small mouths, and the drooling has purpose. The paint on Sophie is food quality.Sophie was born in 1961 in France. The maker had wondered why there was nothing available for babies and made Sophie from the finest rubber, phthalate-free on St Sophie's Day, thus the name was born. Since that time millions of Sophie's populate the world. She is soft and for babies little hands easy to grasp. Violet especially loves the bumpy head and horns of Sophie. Sophie has a long neck that easy to grasp and twist. She has lovely, sizable spots that attract Violet's attention. Sophie has happy little squeaks that bring squeals of delight from Viol

In [28]:
giraffe_reviews[1]['review']

'I\'ll be honest...I bought this toy because all the hip parents seem to have one too and I wanted to be a part of the "hip parent" crowd. The price-tag was somewhat of a deterent but I prevailed and purchased this teether for my daughter.At first, Lily didn\'t know what to make of of Sophie and showed little interest in the polka-dotted creature. I continued to introduce Lily to Sophie and kept the toy in the carrier so that it was on-hand during transitions. Eventually, Lily discovered what a wonderful experience it was to gnaw on the hooves and ears and these two have never been far apart since.Lily really enjoys gumming all the different parts of Sophie like no other teether we have. The size of the toy is great as it is somewhat substantial and so easy for a little one to grasp and hold onto. Lily really enjoys hearing Sophie squeak and will smile whenever Sophie makes a noise or pops her head up from Mommy\'s lap to say hello.People have stopped and commented on Sophie and to the

# Most negative reivews

In [29]:
giraffe_reviews[-1]['review']

"This children's toy is nostalgic and very cute. However, there is a distinct rubber smell and a very odd taste, yes I tried it, that my baby did not enjoy. Also, if it is soiled it is extremely difficult to clean as the rubber is a kind of porus material and does not clean well. The final thing is the squeaking device inside which stopped working after the first couple of days. I returned this item feeling I had overpaid for a toy that was defective and did not meet my expectations. Please do not be swayed by the cute packaging and hype surounding it as I was. One more thing, I was given a full refund from Amazon without any problem."

In [30]:
giraffe_reviews[-2]['review']

'I wanted to love this product and was excited to buy it when I became pregnant but am now hesitant to let my baby use it after reading about the recall in Europe. Apparently, as I understand it, their toxin standards of measurement are lower than ours so they have not been recalled here (apparently we are OK with low levels of nitrates in the toys our children put in their mouths, but Europeans are not...hmmm)...Be that as it may, toxins registering even CLOSE to a dangerous level made me nervous about using. After digging around online I did discover that the company claims to have changed the product after a certain date and lists manufacturing codes so you can check yours (those listed were made after a certain date and are said to be safer). Sadly mine was not made after the &#34;improved&#34; date but I could not return it because there was no formal recall in our country. I considered returning it and hunting for one with an approved manufacturing date but man that was just too 

# Week 3 Questions

In [34]:
selected_words = ['awesome', 'great', 'fantastic', 'amazing', 'love', 'horrible', 'bad', 'terrible', 'awful', 'wow', 'hate']

1. Use .apply() to build a new feature with the counts for each of the selected_words: In the notebook above, we created a column ‘word_count’ with the word counts for each review. Our first task is to create a new column in the products SFrame with the counts for each selected_word above, and, in the process, we will see how the method .apply() can be used to create new columns in our data (our features) and how to use a Python function, which is an extremely useful concept to grasp!

1.1 Our first goal is to create a column products[‘awesome’] where each row contains the number of times the word ‘awesome’ showed up in the review for the corresponding product, and 0 if the review didn’t show up. One way to do this is to look at the each row ‘word_count’ column and follow this logic:

If ‘awesome’ shows up in the word counts for a particular product (row of the products SFrame), then we know how often ‘awesome’ appeared in the review,

if ‘awesome’ doesn’t appear in the word counts, then it didn’t appear in the review, and we should set the count for ‘awesome’ to 0 in this review.

In [93]:
def awesome_count(dict):
    if 'awesome' in dict:
        return dict['awesome']
    else:
        return 0
    
def great_count(dict):
    if 'great' in dict:
        return dict['great']
    else:
        return 0
    
def fantastic_count(dict):
    if 'fantastic' in dict:
        return dict['fantastic']
    else:
        return 0
    
def amazing_count(dict):
    if 'amazing' in dict:
        return dict['amazing']
    else:
        return 0
    
def love_count(dict):
    if 'love' in dict:
        return dict['love']
    else:
        return 0

def horrible_count(dict):
    if 'horrible' in dict:
        return dict['horrible']
    else:
        return 0
    
def bad_count(dict):
    if 'bad' in dict:
        return dict['bad']
    else:
        return 0
    
def terrible_count(dict):
    if 'terrible' in dict:
        return dict['terrible']
    else:
        return 0
    
def awful_count(dict):
    if 'awful' in dict:
        return dict['awful']
    else:
        return 0
    
def wow_count(dict):
    if 'wow' in dict:
        return dict['wow']
    else:
        return 0
    
def hate_count(dict):
    if 'hate' in dict:
        return dict['hate']
    else:
        return 0
    


In [98]:
products1 = turicreate.SFrame('amazon_baby.sframe')
products1['word_count'] = turicreate.text_analytics.count_words(products['review'])

In [99]:
products1['awesome'] = products1['word_count'].apply(awesome_count)
products1['great'] = products1['word_count'].apply(great_count)
products1['fantastic'] = products1['word_count'].apply(fantastic_count)
products1['amazing'] = products1['word_count'].apply(amazing_count)
products1['love'] = products1['word_count'].apply(love_count)
products1['horrible'] = products1['word_count'].apply(horrible_count)
products1['bad'] = products1['word_count'].apply(bad_count)
products1['terrible'] = products1['word_count'].apply(terrible_count)
products1['awful'] = products1['word_count'].apply(awful_count)
products1['wow'] = products1['word_count'].apply(wow_count)
products1['hate'] = products1['word_count'].apply(hate_count)

In [101]:
for word in selected_words:
    print(word)
    print(products1[word].sum())

awesome
4075.0
great
59536.0
fantastic
1765.0
amazing
2726.0
love
43867.0
horrible
1245
bad
4950.0
terrible
1282
awful
753
wow
461
hate
1285


min: wow 461

max: great 59536.0

In [102]:
products1 = products1[products1['rating']!= 3]
products1['sentiment'] = products1['rating'] >= 4

2. Create a new sentiment analysis model using only the selected_words as features: In the Jupyter Notebook above, we used word counts for all words as features for our sentiment classifier. Now, you are just going to use the selected_words:

In [103]:
train_data1,test_data1 = products1.random_split(.8,seed=0)
selected_words_model = turicreate.logistic_classifier.create(train_data1,target='sentiment', features=selected_words, validation_set=test_data1)

Logistic regression:

--------------------------------------------------------

Number of examples          : 133448

Number of classes           : 2

Number of feature columns   : 11

Number of unpacked features : 11

Number of coefficients      : 12

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+--------------+-------------------+---------------------+

| 1         | 2        | 0.067494     | 0.847401          | 0.845874            |

| 2         | 3        | 0.112986     | 0.847514          | 0.846085            |

| 3         | 4        | 0.156442     | 0.847626          | 0.846115            |

| 4         | 5        | 0.199646     | 0.847708          | 0.846385            |

| 5         | 6        | 0.246541     | 0.847708          | 0.846385            |

| 6         | 7        | 0.290394     | 0.847708          | 0.846385            |

+-----------+----------+--------------+-------------------+---------------------+

SUCCESS: Optimal solution found.

In [104]:
selected_words_model

Class                          : LogisticClassifier

Schema
------
Number of coefficients         : 12
Number of examples             : 133448
Number of classes              : 2
Number of feature columns      : 11
Number of unpacked features    : 11

Hyperparameters
---------------
L1 penalty                     : 0.0
L2 penalty                     : 0.01

Training Summary
----------------
Solver                         : newton
Solver iterations              : 6
Solver status                  : SUCCESS: Optimal solution found.
Training time (sec)            : 0.3059

Settings
--------
Log-likelihood                 : 52926.808

Highest Positive Coefficients
-----------------------------
love                           : 1.3593
(intercept)                    : 1.3366
awesome                        : 1.1335
amazing                        : 1.1001
fantastic                      : 0.8858

Lowest Negative Coefficients
----------------------------
horrible                       : -2.2513
ter

In [105]:
coef_table = selected_words_model.coefficients

In [106]:
coef_table = coef_table.sort('value', ascending=False)
coef_table

Columns:
	name	str
	index	str
	class	int
	value	float
	stderr	float

Rows: 12

Data:
+-------------+-------+-------+-----------------------+----------------------+
|     name    | index | class |         value         |        stderr        |
+-------------+-------+-------+-----------------------+----------------------+
|     love    |  None |   1   |   1.3592688669225153  | 0.028068300152099435 |
| (intercept) |  None |   1   |   1.3365913848877558  | 0.008929969787656753 |
|   awesome   |  None |   1   |   1.133534666034145   | 0.08399643983187526  |
|   amazing   |  None |   1   |   1.1000933113660283  | 0.09954776260465983  |
|  fantastic  |  None |   1   |   0.8858047568814295  | 0.11167591293399656  |
|    great    |  None |   1   |   0.8630655001196618  | 0.018955052444377323 |
|     wow     |  None |   1   | -0.009538236067678897 |  0.1604641122471166  |
|     bad     |  None |   1   |  -0.9914778800650565  | 0.03848428664699063  |
|     hate    |  None |   1   |  -1.3484407222463124  | 0.07715698604297333  |
|    awful    |  None |   1   |  -2.0529082040313513  | 0.10099735435259259  |
+-------------+-------+-------+-----------------------+----------------------+
[12 rows x 5 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [107]:
coef_table.tail()

Columns:
	name	str
	index	str
	class	int
	value	float
	stderr	float

Rows: 10

Data:
+-----------+-------+-------+-----------------------+----------------------+
|    name   | index | class |         value         |        stderr        |
+-----------+-------+-------+-----------------------+----------------------+
|  awesome  |  None |   1   |   1.133534666034145   | 0.08399643983187526  |
|  amazing  |  None |   1   |   1.1000933113660283  | 0.09954776260465983  |
| fantastic |  None |   1   |   0.8858047568814295  | 0.11167591293399656  |
|   great   |  None |   1   |   0.8630655001196618  | 0.018955052444377323 |
|    wow    |  None |   1   | -0.009538236067678897 |  0.1604641122471166  |
|    bad    |  None |   1   |  -0.9914778800650565  | 0.03848428664699063  |
|    hate   |  None |   1   |  -1.3484407222463124  | 0.07715698604297333  |
|   awful   |  None |   1   |  -2.0529082040313513  | 0.10099735435259259  |
|  terrible |  None |   1   |   -2.223661436085127  | 0.07731736203785755  |
|  horrible |  None |   1   |   -2.251335236759093  | 0.08020249388788442  |
+-----------+-------+-------+-----------------------+----------------------+
[10 rows x 5 columns]

most positive weight: love

most negative weight: horrible

3. Comparing the accuracy of different sentiment analysis model: Using the method

In [108]:
results1 = selected_words_model.evaluate(test_data1)

In [109]:
results1

{'accuracy': 0.8463848186404036,
 'auc': 0.6936022046674926,
 'confusion_matrix': Columns:
 	target_label	int
 	predicted_label	int
 	count	int
 
 Rows: 4
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |      1       |        0        |  159  |
 |      0       |        0        |  371  |
 |      0       |        1        |  4957 |
 |      1       |        1        | 27817 |
 +--------------+-----------------+-------+
 [4 rows x 3 columns],
 'f1_score': 0.9157860082304526,
 'log_loss': 0.3962265467087378,
 'precision': 0.8487520595594068,
 'recall': 0.9943165570488991,
 'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 100001
 
 Data:
 +-----------+--------------------+-----+-------+------+
 | threshold |        fpr         | tpr |   p   |  n   |
 +-----------+--------------------+-----+-------+------+
 |    0.0    |        1.0         | 1.0 | 27976 | 532

In [121]:
results2 = sentiment_model.evaluate(products1)
results2

{'accuracy': 0.9656795720591057,
 'auc': 0.983638617259335,
 'confusion_matrix': Columns:
 	target_label	int
 	predicted_label	int
 	count	int
 
 Rows: 4
 
 Data:
 +--------------+-----------------+--------+
 | target_label | predicted_label | count  |
 +--------------+-----------------+--------+
 |      1       |        0        |  2569  |
 |      0       |        1        |  3154  |
 |      0       |        0        | 23339  |
 |      1       |        1        | 137690 |
 +--------------+-----------------+--------+
 [4 rows x 3 columns],
 'f1_score': 0.9796409145402218,
 'log_loss': 0.11725006745036073,
 'precision': 0.9776064298088666,
 'recall': 0.9816838848130958,
 'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 100001
 
 Data:
 +-----------+--------------------+--------------------+--------+-------+
 | threshold |        fpr         |        tpr         |   p    |   n   |
 +-----------+--------------------+--------------------+--------+---

accuracy of selected_words_model: 0.8463848186404036

accuracy of sentiment_model: 0.9656795720591057

In [111]:
products1['predicted_sentiment'] = selected_words_model.predict(products1, output_type='probability')

In [112]:
products1.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	int
	bad	float
	terrible	int
	awful	int
	wow	int
	hate	int
	sentiment	int
	predicted_sentiment	float

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
| Baby Tracker&reg; - Daily ... | I love this journal and ou... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+---------+-------+-----------+---------+------+
|           word_count          | awesome | great | fantastic | amazing | love |
+-------------------------------+---------+-------+-----------+---------+------+
| {'recommend': 1.0, 'highly... |   0.0   |  0.0  |    0.0    |   0.0   | 1.0  |
| {'quilt': 1.0, 'of': 1.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'tool': 1.0, 'clever': 1.... |   0.0   |  0.0  |    0.0    |   0.0   | 2.0  |
| {'rock': 1.0, 'many': 1.0,... |   0.0   |  1.0  |    0.0    |   0.0   | 1.0  |
| {'thumb': 1.0, 'or': 1.0, ... |   0.0   |  1.0  |    0.0    |   0.0   | 0.0  |
| {'for': 1.0, 'barnes': 1.0... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'right': 1.0, 'because': ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'like': 1.0, 'and': 1.0, ... |   0.0   |  0.0  |    1.0    |   0.0   | 0.0  |
| {'in': 1.0, 'pages': 1.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'tracker': 1.0, 'now': 1.... |   0.0   |  0.0  |    0.0    |   0.0   | 2.0  |
+-------------------------------+---------+-------+-----------+---------+------+
+----------+-----+----------+-------+-----+------+-----------+
| horrible | bad | terrible | awful | wow | hate | sentiment |
+----------+-----+----------+-------+-----+------+-----------+
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
+----------+-----+----------+-------+-----+------+-----------+
+---------------------+
| predicted_sentiment |
+---------------------+
|  0.936781924479937  |
|  0.7919288370624453 |
|  0.9829620803909053 |
|  0.9723186788855891 |
|  0.9002186948093641 |
|  0.7919288370624453 |
|  0.7919288370624453 |
|  0.9022427410616031 |
|  0.7919288370624453 |
|  0.9829620803909053 |
+---------------------+
[10 rows x 17 columns]

In [123]:
products3 = products1[products1['great']!= 0]
products3

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	int
	bad	float
	terrible	int
	awful	int
	wow	int
	hate	int
	sentiment	int
	predicted_sentiment	float

Rows: Unknown

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| Nature's Lullabies Second ... | I had a hard time finding ... |  5.0   |
|  Lamaze Peekaboo, I Love You  | I like how the book has a ... |  5.0   |
|  Lamaze Peekaboo, I Love You  | Beautiful little book.  A ... |  5.0   |
|  Lamaze Peekaboo, I Love You  | What a great book for babi... |  5.0   |
| SoftPlay Peek-A-Boo Where'... | My baby absolutely loves E... |  5.0   |
| SoftPlay Peek-A-Boo Where'... | I bought two for recent ba... |  5.0   |
| SoftPlay Peek-A-Boo Where'... | Very cute interactive book... |  5.0   |
| SoftPlay 4 Pack Block Book... | I have teething babies and... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+---------+-------+-----------+---------+------+
|           word_count          | awesome | great | fantastic | amazing | love |
+-------------------------------+---------+-------+-----------+---------+------+
| {'rock': 1.0, 'many': 1.0,... |   0.0   |  1.0  |    0.0    |   0.0   | 1.0  |
| {'thumb': 1.0, 'or': 1.0, ... |   0.0   |  1.0  |    0.0    |   0.0   | 0.0  |
| {'reference': 1.0, 'have':... |   0.0   |  2.0  |    0.0    |   0.0   | 0.0  |
| {'family': 1.0, 'over': 1.... |   0.0   |  1.0  |    0.0    |   0.0   | 0.0  |
| {'interacting': 1.0, 'when... |   0.0   |  1.0  |    0.0    |   0.0   | 1.0  |
| {'baby': 1.0, '10': 1.0, '... |   0.0   |  1.0  |    0.0    |   0.0   | 0.0  |
| {'a': 1.0, 'amazon': 1.0, ... |   0.0   |  2.0  |    0.0    |   0.0   | 0.0  |
| {'elmo': 1.0, 'love': 1.0,... |   0.0   |  1.0  |    0.0    |   0.0   | 1.0  |
| {'babies': 1.0, 'great': 1... |   0.0   |  1.0  |    0.0    |   0.0   | 0.0  |
| {'bank': 1.0, 'didn': 1.0,... |   0.0   |  3.0  |    0.0    |   0.0   | 0.0  |
+-------------------------------+---------+-------+-----------+---------+------+
+----------+-----+----------+-------+-----+------+-----------+
| horrible | bad | terrible | awful | wow | hate | sentiment |
+----------+-----+----------+-------+-----+------+-----------+
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
+----------+-----+----------+-------+-----+------+-----------+
+---------------------+
| predicted_sentiment |
+---------------------+
|  0.9723186788855891 |
|  0.9002186948093641 |
|  0.9553286211620002 |
|  0.9002186948093641 |
|  0.9723186788855891 |
|  0.9002186948093641 |
|  0.9553286211620002 |
|  0.9723186788855891 |
|  0.9002186948093641 |
|  0.9806550219614182 |
+---------------------+
[? rows x 17 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

In [126]:
products3['predicted_sentiment'].sum()/len(products3)

0.9215489664427361

4. Interpreting the difference in performance between the models: To understand why the model with all word counts performs better than the one with only the selected_words, we will now examine the reviews for a particular product.

In [113]:
diaper_champ_reviews = products1[products1['name']== 'Baby Trend Diaper Champ']
diaper_champ_reviews

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	int
	bad	float
	terrible	int
	awful	int
	wow	int
	hate	int
	sentiment	int
	predicted_sentiment	float

Rows: Unknown

Data:
+-------------------------+-------------------------------+--------+
|           name          |             review            | rating |
+-------------------------+-------------------------------+--------+
| Baby Trend Diaper Champ | Ok - newsflash.  Diapers a... |  4.0   |
| Baby Trend Diaper Champ | My husband and I selected ... |  1.0   |
| Baby Trend Diaper Champ | Excellent diaper disposal ... |  5.0   |
| Baby Trend Diaper Champ | We love our diaper champ. ... |  5.0   |
| Baby Trend Diaper Champ | Two girlfriends and two fa... |  5.0   |
| Baby Trend Diaper Champ | I waited to review this un... |  4.0   |
| Baby Trend Diaper Champ | I have had a diaper genie ... |  1.0   |
| Baby Trend Diaper Champ | I originally put this item... |  5.0   |
| Baby Trend Diaper Champ | I am so glad I got the Dia... |  5.0   |
| Baby Trend Diaper Champ | We had 2 diaper Genie's bo... |  4.0   |
+-------------------------+-------------------------------+--------+
+-------------------------------+---------+-------+-----------+---------+------+
|           word_count          | awesome | great | fantastic | amazing | love |
+-------------------------------+---------+-------+-----------+---------+------+
| {'convenient': 1.0, 'more'... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'system': 1.0, 'try': 1.0... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'nose': 1.0, 'for': 2.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'out': 1.0, 'pull': 1.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 1.0  |
| {'winter': 1.0, 'outside':... |   0.0   |  0.0  |    0.0    |   1.0   | 0.0  |
| {'mom': 1.0, 'my': 1.0, 'e... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'yuck': 1.0, 'clean': 1.0... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'price': 1.0, 'suggestion... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'best': 1.0, 'that': 1.0,... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'no': 1.0, 'regular': 1.0... |   0.0   |  0.0  |    0.0    |   0.0   | 2.0  |
+-------------------------------+---------+-------+-----------+---------+------+
+----------+-----+----------+-------+-----+------+-----------+
| horrible | bad | terrible | awful | wow | hate | sentiment |
+----------+-----+----------+-------+-----+------+-----------+
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     0     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    1     | 0.0 |    0     |   0   |  1  |  0   |     1     |
|    0     | 1.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     0     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
+----------+-----+----------+-------+-----+------+-----------+
+---------------------+
| predicted_sentiment |
+---------------------+
|  0.7919288370624453 |
|  0.7919288370624453 |
|  0.7919288370624453 |
|  0.936781924479937  |
|  0.5438399411170777 |
|  0.5854321171706491 |
|  0.7919288370624453 |
|  0.7919288370624453 |
|  0.7919288370624453 |
|  0.9829620803909053 |
+---------------------+
[? rows x 17 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

In [115]:
diaper_champ_reviews = diaper_champ_reviews.sort('predicted_sentiment', ascending=False)
diaper_champ_reviews

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	int
	bad	float
	terrible	int
	awful	int
	wow	int
	hate	int
	sentiment	int
	predicted_sentiment	float

Rows: 298

Data:
+-------------------------+-------------------------------+--------+
|           name          |             review            | rating |
+-------------------------+-------------------------------+--------+
| Baby Trend Diaper Champ | I LOVE LOVE LOVE this prod... |  4.0   |
| Baby Trend Diaper Champ | I received my Diaper Champ... |  5.0   |
| Baby Trend Diaper Champ | Love it, love it, love it!... |  5.0   |
| Baby Trend Diaper Champ | Let me just say, I LOVE TH... |  5.0   |
| Baby Trend Diaper Champ | I have been using this dia... |  5.0   |
| Baby Trend Diaper Champ | Works great - no smells.  ... |  5.0   |
| Baby Trend Diaper Champ | I love this diaper pale an... |  5.0   |
| Baby Trend Diaper Champ | We had 2 diaper Genie's bo... |  4.0   |
| Baby Trend Diaper Champ | Love the Diaper Champ.  I ... |  4.0   |
| Baby Trend Diaper Champ | I have a 10 year old daugh... |  5.0   |
+-------------------------+-------------------------------+--------+
+-------------------------------+---------+-------+-----------+---------+------+
|           word_count          | awesome | great | fantastic | amazing | love |
+-------------------------------+---------+-------+-----------+---------+------+
| {'sorry': 1.0, 'be': 1.0, ... |   0.0   |  1.0  |    0.0    |   0.0   | 3.0  |
| {'second': 2.0, 'have': 1.... |   0.0   |  0.0  |    0.0    |   0.0   | 3.0  |
| {'parents': 1.0, 'new': 1.... |   0.0   |  0.0  |    0.0    |   0.0   | 3.0  |
| {'friends': 1.0, 'feces': ... |   0.0   |  1.0  |    1.0    |   0.0   | 1.0  |
| {'recommend': 1.0, 'would'... |   0.0   |  2.0  |    0.0    |   0.0   | 1.0  |
| {'plastic': 1.0, 'in': 1.0... |   0.0   |  2.0  |    0.0    |   0.0   | 1.0  |
| {'easy': 1.0, 'how': 1.0, ... |   0.0   |  2.0  |    0.0    |   0.0   | 1.0  |
| {'no': 1.0, 'regular': 1.0... |   0.0   |  0.0  |    0.0    |   0.0   | 2.0  |
| {'want': 1.0, 'wouldn': 1.... |   0.0   |  0.0  |    0.0    |   0.0   | 2.0  |
| {'sorry': 1.0, 'be': 1.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 2.0  |
+-------------------------------+---------+-------+-----------+---------+------+
+----------+-----+----------+-------+-----+------+-----------+
| horrible | bad | terrible | awful | wow | hate | sentiment |
+----------+-----+----------+-------+-----+------+-----------+
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
+----------+-----+----------+-------+-----+------+-----------+
+---------------------+
| predicted_sentiment |
+---------------------+
|  0.9981253623335122 |
|  0.9955677154227358 |
|  0.9955677154227358 |
|  0.9883959649944544 |
|  0.9881322423088146 |
|  0.9881322423088146 |
|  0.9881322423088146 |
|  0.9829620803909053 |
|  0.9829620803909053 |
|  0.9829620803909053 |
+---------------------+
[298 rows x 17 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [116]:
diaper_champ_reviews[0]['review']

'I LOVE LOVE LOVE this product! It is SO much easier to use than the Diaper Genie, (you need a PHD in poopy to figure out how to use the darn thing!) and it even takes the same bags as my kitchen trash can, shich is super convenient, and cost efficient as I can buy them in bulk.The only reason for not rating it a 5 star was that I did have one small problem with it. The foam gasket in the barrell which keeps the poopy smell inside the unit ripped somehow, and it got VERY stinky. HOWEVER, I contacted the manufacturer though their website, and received an email back the same day stating that this was unusual, and that replacement gaskets were on their way to me. They arrived inside of a week and after replacing, it works great again! (They even sent me extras should it happen again)I HIGHLY reccomend this diaper pail over ANY competitors, you will not be sorry!'

In [117]:
diaper_champ_reviews[-1]['review']

"......all I can say is the smell is horrible.....1 star..... Please don't buy this one!"

In [118]:
selected_words_model.predict(diaper_champ_reviews[0:1], output_type='probability')

dtype: float
Rows: 1
[0.9981253623335122]

In [119]:
sentiment_model.predict(diaper_champ_reviews[0:1], output_type='probability')

dtype: float
Rows: 1
[0.9999819570853404]